pcap-linux.c

```diff
@@ -1469,42 +1474,60 @@ pcap_activate_linux(pcap_t *handle)
 		 * Success.
 		 * Try to use memory-mapped access.
 		 */
-		switch (activate_mmap(handle, &status)) {
+		ret = activate_mmap(handle, &status);
+		if (ret == -1) {
+			/*
+			 * We failed to set up to use it, or the
+			 * kernel supports it, but we failed to
+			 * enable it.  status has been set to the
+			 * error status to return and, if it's
+			 * PCAP_ERROR, handle->errbuf contains
+			 * the error message.
+			 */
+			goto fail;
+		}
 
-		case 1:
+
+		if (ret == 1) {
 			/*
 			 * We succeeded.  status has been
 			 * set to the status to return,
 			 * which might be 0, or might be
 			 * a PCAP_WARNING_ value.
-			 *
-			 * Set the timeout to use in poll() before
-			 * returning.
 			 */
-			set_poll_timeout(handlep);
-			return status;
-
-		case 0:
 			/*
-			 * Kernel doesn't support it - just continue
-			 * with non-memory-mapped access.
+			 * Now that we have activated the mmap ring, we can
+			 * set the correct protocol.
```

Comment:
"...we can bind to the specified interface, if one was specified, and set the correct protocol" - that call does both.

Reply:
Yes, you are right.

The problem that I had, is that I tried to make my match as minimal as possible, and not to change the overall structure of the code too much.  This means that iface_bind often gets called twice, once with (ifindex, 0) and then again with (ifindex, protocol).

I assumed that binding twice would be best - if the ifindex is bad, or the device has errors, we get those on the first bind call, and the code flow is unchanged by my patch.  The second bind call is less likely to fail, as "less is changing" (I felt, at least).